src/par_iter/internal.rs

```diff
@@ -95,6 +111,22 @@ pub trait UnindexedConsumer<ITEM>: Consumer<ITEM> {
 pub trait UnindexedProducer: IntoIterator + Send + Sized {
     fn can_split(&self) -> bool;
     fn split(self) -> (Self, Self);
+
+    /// Iterate the producer, feeding each element to `folder`, and
+    /// stop when the folder is full (or all elements have been consumed).
+    ///
+    /// The provided implementation is sufficient for most iterables.
+    fn fold_with<F>(self, mut folder: F) -> F
+        where F: Folder<Self::Item>,
+    {
+        for item in self {
+            folder = folder.consume(item);
```

Comment:
> At some point, we discussed whether we should have Producer: UnindexedProducer, but we didn't have a good reason for it (even though it logically makes sense). The existence of this duplicate code might provide such a reason.

Yes, although IIRC we were also wanting `impl<T: Producer> UnindexedProducer for T` to avoid duplicating `can_split`/`split` boilerplate, but that would really want specialization.  Plus that boilerplate would never actually be invoked.

In the case of just `fold_while`, perhaps we could make a `BaseProducer` common to both, so the boilerplate is a one-line empty `impl` in most cases.  Or we could extract this body to a shared free function just to avoid the duplication.  But this is only 6 lines, let's not overthink it. :)